### ISSUE TITLE
[Bug] trtllm_fp4_block_scale_routed_moe: fallback tactic returns non-finite outputs with per_token_scale at large num_tokens (SM103)

### ISSUE THREAD

**Issue by sshleifer** [flashinfer-ai/flashinfer#4588]
## Summary

`trtllm_fp4_block_scale_routed_moe` with `per_token_scale` (per-token NVFP4 activation scaling) produces **non-finite outputs** when it runs on the fallback tactic (tactic `-1`, i.e. no autotune cache entry for the bucket) and `num_tokens` is large. With identical weights, routing, and inputs:

| config | result |
|---|---|
| fallback tactic, `per_token_scale`, num_tokens=8064 | **~0.7–13% of output elements NaN/Inf** (count varies run to run) |
| fallback tactic, static activation scale (`per_token_scale=None`), same everything | clean, cosine 0.973 vs bf16 reference |
| autotuned tactic, `per_token_scale`, same everything | clean, cosine 0.973 vs bf16 reference |

An untuned tactic may be slow, but should never be wrong. In serving this shows up as `NaN detected in logits` on real prefill batches whenever the extend-shaped warmup pass didn't run.

Additional observations:

- `num_tokens <= 5120` is clean; `>= 6144` is broken (threshold between 5120 and 6144 at these shapes).
- `enable_pdl` True/False does not matter.
- The non-finite count varies run to run at fixed seed, and grows to **~60–80%** of the output if free GPU memory is pre-filled with NaN bytes before the call — consistent with the fallback config reading memory it shouldn't (uninitialized workspace or an out-of-bounds/garbage read of the per-token scales).
- Quantizer backend for the activations (`cuda` vs `cute-dsl`) does not matter; routing skew does not matter.

Possibly related: #4190 reports the *perf* side of the same regime (default tactic at >=8k tokens on this op, autotuned tactic fine); this issue is the correctness side with `per_token_scale`.

## Environment

- flashinfer 0.6.16.post4 (also reproduced on 0.6.17): `flashinfer-python` + `flashinfer-cubin` wheels
- 1× NVIDIA GB300 (SM103), driver 580.159.03
- torch 2.11.0+cu130, CUDA 13.0, aarch64

## Repro

Single GPU, only torch + flashinfer. Shapes are a production MoE per-TP-rank slice: hidden=6144, intermediate=768, 256 experts, top_k=6, gated silu; num_tokens=8064 is a real chunked-prefill batch.

```
$ python repro_trtllm_fp4_routed_moe_nan.py
UNTUNED per-token scales: non-finite = 4865152/49545216, cosine vs bf16 ref = nan
UNTUNED static act scale: non-finite = 0/49545216, cosine vs bf16 ref = 0.9730
TUNED   per-token scales: non-finite = 0/49545216, cosine vs bf16 ref = 0.9731
```

<details>
<summary>repro_trtllm_fp4_routed_moe_nan.py</summary>

```python
"""trtllm_fp4_block_scale_routed_moe: the untuned fallback tactic produces
non-finite outputs when per_token_scale is used and num_tokens is large (>= ~6k).

Single GPU (tested on GB300 / SM103). Depends only on torch + flashinfer.

Shapes are the per-TP-rank MoE of a production model (hidden=6144,
intermediate=768, 256 experts, top_k=6, gated silu). num_tokens=8064 is a real
chunked-prefill batch size.

Expected output (buggy):
    UNTUNED per-token scales      : non-finite = <large>/49545216   <-- BUG
    UNTUNED static act scale      : non-finite = 0/49545216, cos ~0.97
    TUNED   per-token scales      : non-finite = 0/49545216, cos ~0.97

i.e. with identical weights/routing/inputs, only the (fallback tactic,
per_token_scale) combination is broken; the autotuned tactic and the
static-activation-scale fallback are both fine. The non-finite count varies
run to run. enable_pdl True/False does not matter. num_tokens <= 5120 is
clean; >= 6144 is not.
"""

import torch

from flashinfer import SfLayout, nvfp4_block_scale_interleave, nvfp4_quantize
from flashinfer.autotuner import autotune
from flashinfer.fused_moe import trtllm_fp4_block_scale_routed_moe
from flashinfer.fused_moe.core import (
    _maybe_get_cached_w3_w1_permute_indices,
    get_w2_permute_indices_with_cache,
)

torch.manual_seed(0)
DEV = "cuda"
E, D, F, TOPK = 256, 6144, 768, 6
T = 8064
FP4_MAX, E4M3_MAX = 6.0, 448.0
EPILOGUE_TILE_M = 128


def quant_weight(w: torch.Tensor):
    """(rows, D) bf16 -> (packed fp4 uint8, linear e4m3 block sf, global dequant scale)."""
    gs = (FP4_MAX * E4M3_MAX) / w.abs().amax().float()
    data, sf = nvfp4_quantize(w, gs.reshape(1), sfLayout=SfLayout.layout_linear, do_shuffle=False)
    return data, sf, 1.0 / gs


def main() -> None:
    # Random experts: rows [0:F] = up (w3), rows [F:2F] = gate (w1).
    w13 = torch.randn(E, 2 * F, D, dtype=torch.bfloat16, device=DEV) * 0.02
    w2 = torch.randn(E, D, F, dtype=torch.bfloat16, device=DEV) * 0.02

    w13_q = torch.empty(E, 2 * F, D // 2, dtype=torch.uint8, device=DEV)
    w13_sf = torch.empty(E, 2 * F, D // 16, dtype=torch.uint8, device=DEV)
    w2_q = torch.empty(E, D, F // 2, dtype=torch.uint8, device=DEV)
    w2_sf = torch.empty(E, D, F // 16, dtype=torch.uint8, device=DEV)
    g1_alphas = torch.empty(E, dtype=torch.float32, device=DEV)
    g2_alphas = torch.empty(E, dtype=torch.float32, device=DEV)
    for e in range(E):
        w13_q[e], sf, g1_alphas[e] = quant_weight(w13[e])
        w13_sf[e] = sf.view(torch.uint8).reshape(2 * F, D // 16)
        w2_q[e], sf, g2_alphas[e] = quant_weight(w2[e])
        w2_sf[e] = sf.view(torch.uint8).reshape(D, F // 16)

    # Shuffle weights + interleave block scales for the trtllm-gen kernel.
    cache: dict = {}
    w13_shuf = torch.empty_like(w13_q)
    w2_shuf = torch.empty_like(w2_q)
    w13_sf_shuf = torch.empty(E, 2 * F * (D // 16), dtype=torch.uint8, device=DEV)
    w2_sf_shuf = torch.empty(E, D * (F // 16), dtype=torch.uint8, device=DEV)
    for e in range(E):
        pi = _maybe_get_cached_w3_w1_permute_indices(
            cache, w13_q[e], EPILOGUE_TILE_M, is_gated_act_gemm=True
        )
        w13_shuf[e] = w13_q[e][pi.to(DEV)]
        pi = _maybe_get_cached_w3_w1_permute_indices(
            cache, w13_sf[e], EPILOGUE_TILE_M, num_elts_per_sf=16, is_gated_act_gemm=True
        )
        w13_sf_shuf[e] = nvfp4_block_scale_interleave(w13_sf[e][pi.to(DEV)].contiguous()).flatten()
        pi = get_w2_permute_indices_with_cache(cache, w2_q[e], EPILOGUE_TILE_M)
        w2_shuf[e] = w2_q[e][pi.to(DEV)]
        pi = get_w2_permute_indices_with_cache(
            cache, w2_sf[e], EPILOGUE_TILE_M, num_elts_per_sf=16
        )
        w2_sf_shuf[e] = nvfp4_block_scale_interleave(w2_sf[e][pi.to(DEV)].contiguous()).flatten()
    w13_sf_shuf = w13_sf_shuf.view(torch.float8_e4m3fn).reshape(E, 2 * F, D // 16)
    w2_sf_shuf = w2_sf_shuf.view(torch.float8_e4m3fn).reshape(E, D, F // 16)

    x = torch.randn(T, D, dtype=torch.bfloat16, device=DEV)

    # Routing: random experts, softmax weights, packed as (expert_id << 16) | bf16 weight bits.
    topk_ids = torch.stack([torch.randperm(E, device=DEV)[:TOPK] for _ in range(T)]).int()
    topk_weights = torch.softmax(torch.randn(T, TOPK, device=DEV), dim=-1)
    packed = (topk_ids.to(torch.int32) << 16) | (
        topk_weights.to(torch.bfloat16).view(torch.int16).to(torch.int32) & 0xFFFF
    )

    # bf16 reference from the unquantized tensors: sum_k w_k * (silu(gate) * up) @ w2.
    ref = torch.zeros(T, D, dtype=torch.float32, device=DEV)
    for k in range(TOPK):
        we = topk_ids[:, k].long()
        up = torch.einsum("td,tfd->tf", x, w13[we, :F])
        gate = torch.einsum("td,tfd->tf", x, w13[we, F:])
        h = torch.nn.functional.silu(gate) * up
        ref += topk_weights[:, k, None] * torch.einsum("tf,tdf->td", h, w2[we]).float()

    def quant_act(per_token: bool):
        if per_token:
            gs_inv = torch.full(
                (1,), 1.0 / (E4M3_MAX * FP4_MAX), dtype=torch.float32, device=DEV
            )
            x_q, x_sf, pts = nvfp4_quantize(
                x, gs_inv, sfLayout=SfLayout.layout_linear, per_token_activation=True
            )
            # per-token path: activation global scale is folded into per_token_scale
            o1c, o1g, o2 = g1_alphas, g1_alphas, g2_alphas
        else:
            act_gs = ((E4M3_MAX * FP4_MAX) / x.abs().amax().float()).reshape(1)
            x_q, x_sf = nvfp4_quantize(x, act_gs, sfLayout=SfLayout.layout_linear)
            pts = None
            o1g = (g1_alphas / act_gs).contiguous()
            o1c, o2 = o1g.clone(), g2_alphas
        return x_q.reshape(T, D // 2), x_sf.view(torch.float8_e4m3fn).reshape(T, D // 16), pts, o1c, o1g, o2

    def run(per_token: bool) -> torch.Tensor:
        x_q, x_sf, pts, o1c, o1g, o2 = quant_act(per_token)
        return trtllm_fp4_block_scale_routed_moe(
            topk_ids=packed,
            routing_bias=None,
            hidden_states=x_q,
            hidden_states_scale=x_sf,
            gemm1_weights=w13_shuf,
            gemm1_weights_scale=w13_sf_shuf,
            gemm1_bias=None,
            gemm1_alpha=None,
            gemm1_beta=None,
            gemm1_clamp_limit=None,
            gemm2_weights=w2_shuf,
            gemm2_weights_scale=w2_sf_shuf,
            gemm2_bias=None,
            output1_scale_scalar=o1c,
            output1_scale_gate_scalar=o1g,
            output2_scale_scalar=o2,
            per_token_scale=pts,
            num_experts=E,
            top_k=TOPK,
            n_group=0,
            topk_group=0,
            intermediate_size=F,
            local_expert_offset=0,
            local_num_experts=E,
            routed_scaling_factor=None,
            routing_method_type=1,
            do_finalize=True,
            tune_max_num_tokens=8192,
        )[0]

    def report(name: str, out: torch.Tensor) -> None:
        bad = (~torch.isfinite(out)).sum().item()
        cos = torch.nn.functional.cosine_similarity(
            out.float().flatten(), ref.flatten(), dim=0
        ).item()
        print(f"{name}: non-finite = {bad}/{out.numel()}, cosine vs bf16 ref = {cos:.4f}")

    report("UNTUNED per-token scales", run(per_token=True))  # <-- BUG: non-finite
    report("UNTUNED static act scale", run(per_token=False))  # control: clean
    with torch.inference_mode(), autotune(True):
        run(per_token=True)
    report("TUNED   per-token scales", run(per_token=True))  # clean


if __name__ == "__main__":
    main()
```

</details>

-Robot

**Comment by aleozlx** [comment]
looks like the same as https://github.com/flashinfer-ai/flashinfer/issues/4486

kindly let us know
